backpack-common/src/main/java/net/skyscanner/backpack/calendar2/data/CalendarCell.kt

```diff
@@ -98,40 +100,103 @@ internal fun CalendarCellDay(
-    outOfRange = date !in params.range,
-    contentDescription = date.format(params.dateContentDescriptionFormatter),
-    text = buildSpannedString {
-        val span = TtsSpan.DateBuilder()
-            .setDay(date.dayOfMonth)
-            .setMonth(date.month.ordinal)
-            .setYear(date.year)
-            .setWeekday(date.dayOfWeek.value)
-            .build()
-        append(date.dayOfMonth.toString(), span, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
-    },
-    selection = when (selection) {
-        is CalendarSelection.None -> null
-        is CalendarSelection.Single -> when (date) {
-            selection.date -> CalendarCell.Selection.Single
+): CalendarCell.Day {
+    return CalendarCell.Day(
+        date = date,
+        yearMonth = yearMonth,
+        info = params.cellsInfo[date] ?: CellInfo.Default,
+        outOfRange = date !in params.range,
+        contentDescription = date.format(params.dateContentDescriptionFormatter),
+        stateDescription = stateDescription(date, params.selectionMode, selection),
+        onClickLabel = onClickLabel(date, params.selectionMode, selection),
+        text = buildSpannedString {
+            val span = TtsSpan.DateBuilder()
+                .setDay(date.dayOfMonth)
+                .setMonth(date.month.ordinal)
+                .setYear(date.year)
+                .setWeekday(date.dayOfWeek.value)
+                .build()
+            append(date.dayOfMonth.toString(), span, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
+        },
+        selection = when (selection) {
+            is CalendarSelection.None -> null
+            is CalendarSelection.Single -> when (date) {
+                selection.date -> CalendarCell.Selection.Single
+                else -> null
+            }
+
+            is CalendarSelection.Dates -> when {
+                selection.start == date && selection.end == date -> CalendarCell.Selection.Double
+                selection.start == date && selection.end == null -> CalendarCell.Selection.Single
+                selection.start == date && selection.end != null -> CalendarCell.Selection.Start
+                selection.end == date -> CalendarCell.Selection.End
+                selection.end != null && date in selection -> CalendarCell.Selection.Middle
+                else -> null
+            }
+
+            is CalendarSelection.Month -> when {
+                selection.start == date -> CalendarCell.Selection.StartMonth
+                selection.end == date -> CalendarCell.Selection.EndMonth
+                date in selection -> CalendarCell.Selection.Middle
+                else -> null
+            }
+        },
+    )
+}
+
+private fun stateDescription(
+    date: LocalDate,
+    selectionMode: CalendarParams.SelectionMode,
+    selection: CalendarSelection,
+): String? {
+    return when (selectionMode) {
+        is CalendarParams.SelectionMode.Single -> when (selection) {
+            CalendarSelection.None -> selectionMode.noSelectionState
```

Comment:
Can you please add `is` operator here as well? This way, if the code gets converted from a singleton object to a data class, we won't have to update all the when statements.

Reply:
Thanks, done!